### DISCUSSION TITLE
Does jolt support Static and dynamic friction?

### DISCUSSION THREAD

**Discussion by yosoyfreeman** [godot-jolt/godot-jolt#585] [General]
Does jolt support Static and dynamic friction?, if so, it may be possible to support it in godot jolt at some point? Godot jolt works incredibly well but godot approach to friction is not so predictable. Fine tuning is not possible due to not differentiation between static and dynamic friction. Also, in quite a lot of cases you need friction values extremely high. It is supposed to be from 0 to 1, but we are needing about 100 friction for our rolling ball controller to keep traction on the floor and still slide quite a lot.  I would like what are your plans on this or how limited you are by Godot implementation. Because i'm finding Jolt doing an incredible job but limited by this things.

Thank you for you time,
Freeman.

**Comment by mihe** [comment]
> Does jolt support Static and dynamic friction?

@jrouwe can probably give you a more informed answer here with regards to Jolt's capabilities, but from what I understand this is not a distinction that Jolt makes currently. I guess friction is essentially always dynamic friction?

> It is supposed to be from 0 to 1, but we are needing about 100 friction for our rolling ball controller to keep traction on the floor and still slide quite a lot.

Keep in mind that friction between to objects/surfaces is combined like this:

https://github.com/godot-jolt/godot-jolt/blob/afbcae0ebfb4e14bfd5bb6c8c17b9f38a98b7269/src/spaces/jolt_space_3d.cpp#L63

So regardless of how high you set the friction you will end up with the lowest value of the two surfaces. However, if you enable the "Rough" property on the physics material of your ball it will get a negative friction value, and thereby effectively override the combined friction to always be that of the ball's. You might want to give that a try if you haven't already.

With that said, I'm not sure if it's reasonable to expect an object to absolutely never slide, even if the combined friction is >= 1, but maybe @jrouwe can shed some light on this.

**Reply by jrouwe** [reply]
> @jrouwe can probably give you a more informed answer here with regards to Jolt's capabilities, but from what I understand this is not a distinction that Jolt makes currently. I guess friction is essentially always dynamic friction?

No I did indeed not make a distinction. The current friction value is the 'dynamic' friction. It would not be terribly hard to add support for static friction, but it is hard to spot the difference in normal game scenarios, so I didn't bother. For a ball I would not expect the distinction between static and dynamic friction to do much as a ball needs very little force to make it roll and static friction has a higher max force than dynamic friction but other than that it is the same. 

From the samples:

https://github.com/godot-jolt/godot-jolt/assets/1621693/467764f9-6f2c-435b-a86a-1e1f981ef892

The first ball has a friction of 0, the last 1. The first ball only slides because it has no friction, the rest of the balls behave nearly identical even though they have very different friction values (the blocks, which have the same friction as the balls, do not however).

B.t.w. I don't think godot physics has a distinction between static/dynamic either?

> With that said, I'm not sure if it's reasonable to expect an object to absolutely never slide, even if the combined friction is >= 1, but maybe @jrouwe can shed some light on this.

A friction of 1 means that the friction force can at most be the same as the non-penetration force (= force between ground and object along the surface normal). So if the slope on which the object lies is sufficiently large it will still slide.

**Reply by mihe** [reply]
> B.t.w. I don't think godot physics has a distinction between static/dynamic either?

Correct.

**Comment by yosoyfreeman** [comment]
Thank you both for the clarification. The question was born because while i was studying physics i end up with a lot of videos explaining how static friction makes a big part of keeping things from sliding of ramps and the tutorials i found were in unity where this properties exist. That's why i was curious if we would eventually get this distinction.  For example, in the sphere case you shown, in fact they all behave really similar, making it hard to achieve more specific results. In the case of a sphere, the dynamic friction is always relatively small, but its static friction, which is a constant, would make the most part of fine tuning the movement. Also, as it scales with the normal force and therefore the mass, it can play an important role in some cases.   So is for cases like this that i think it can be a real difference to have at least in physics based games. 

Thank you again for taking the time to answer!

**Reply by jrouwe** [reply]
>  In the case of a sphere, the dynamic friction is always relatively small, but its static friction, which is a constant, would make the most part of fine tuning the movement. Also, as it scales with the normal force and therefore the mass, it can play an important role in some cases.

I don't think this is correct. Both the [static and the dynamic friction depend on the normal force](https://www.geeksforgeeks.org/difference-between-static-friction-and-dynamic-friction/). The only difference is that the static friction coefficient is much higher, but it is only applied when the bodies are not moving relative to each other. Friction is implemented by calculating the force that is needed to reduce the (contact point) velocity difference of 2 objects to zero in one time step. The maximum friction force that can be applied is F = friction coefficient * normal force, so if the calculated force exceeds F then it will be clamped to F. In this equation you would set friction coefficient = static friction coefficient if the point velocity difference between the 2 objects is currently nearly zero and you would set it to dynamic friction coefficient if it is not. So if the calculated force is smaller than the dynamic friction coefficient * normal force, it will not get clamped anyway and there is no difference between static and dynamic friction.

In the samples app you can see this effect, now the friction coefficients go from 0 to 10000 instead of from 0 to 1. You can see that all but the first box (which has no friction) will now stop immediately, but the spheres behave exactly as before (simply because the dynamic friction force was large enough to not trigger clamping of the calculated force):

https://github.com/godot-jolt/godot-jolt/assets/1621693/964be100-5245-49f1-bd62-a06b1471fcc7

So in this example it wouldn't have mattered if the engine supports static friction or not, the dynamic friction coefficient is sufficient.

In Jolt the responsible code is here:

https://github.com/jrouwe/JoltPhysics/blob/1b21180c67930f5669750a7912c55d90407586ee/Jolt/Physics/Constraints/ContactConstraintManager.cpp#L1501-L1523

and the only thing that would change if static friction is implemented is that there would be a secondary friction coefficient mCombinedStaticFriction which would be used to calculate max_lambda_f when the relative velocity at the contact point is near zero.

I think this could even be implemented without changing the library as the friction coefficient can be overridden in the contact callback, so you can do the check if the relative contact point velocity is near zero there and select a static/dynamic friction coefficient based on the result of this check.
